### ISSUE TITLE
Provide a means to embed one library into another

### ISSUE THREAD

**Issue by mikkelfj** [mesonbuild/meson#3349]
Building a staging library with separate compile options can be included in a parent library project using link_with or link_whole, but from inspecting build.ninja it is clear that all the libraries are linked with the final user application and not added to the library. To fix this problem it is necessary to use extract_objects. This has the effect of bypassing the staging library and use the object files directly in the parent, assuming the parent is given the `objects` argument.

This issue is a follow up to https://github.com/mesonbuild/meson/issues/3345
which deals with the same problem but focusing on the part where compiler flags are being set. This could for some use cases be handled by the SIMD module, but it still mostly likely doesn't solve the linker issue, hence this issue.

**Comment by jpakkane** [comment]
What is the actual problem here? I can see that it would break if you install a static library because then trying to use it would fail as knowledge that the temporary static dir must be used. However building executables and shared libraries should work in all cases. What specifically breaks for you?

**Comment by mikkelfj** [comment]
If meson is not used end to end, or a certain unix install convention is not used, the all those tiny libraries need to be carried on to dependent projects rather than just building a self-contained library, agnostic to build systems.

Put in another way, want to be able build a minimum dependency library, with at most a clib dependency, if that.

**Comment by zackw** [comment]
I would expect `link_whole: libA` to embed library A, and I would expect `link_with: libA` not to do so, and in my testing that's exactly how it works.  For instance

```
project('demo', 'c')
libA = static_library('A', sources: ['a.c'])
libB = both_libraries('B', sources: ['b.c'], link_with: libA)
libC = both_libraries('C': sources: ['c.c'], link_whole: libA)
```

embeds libA into libC but not libB when I try it (with 0.53).  @mikkelfj Can you tell us if you see the same thing?  If you do, can you try to figure out what the difference is between your meson.build and this?

**Comment by eli-schwartz** [comment]
I think this is basically just another instance of #3937, and is fixed by #5936

So I think the reason this works for you, @zackw, is because the fix was added to meson in between the time this ticket was posted and the time you commented. :D 

@mikkelfj Closing, because this should work on meson >=0.52.0, but please feel free to re-open if you still have this problem.
